File: src/client_action/src/custom_service.cpp
/**
  * \file custom_service.cpp 
  * \brief Node implementing a custom service.
  * 
  *
  * \details
  *
  * Subscribes to: <BR>
  *   ° /reaching_goal/status
  *
  *
  * Services : <BR>
  * ° /my_custom_server_topic
  *
  * Description :
  * 
  * This node provides a service "/my_custom_server_topic" that prints the number of goals reached and canceled when the service is called.
  * to acheive this it subscribes to the topic '/reaching_goal/status' and increaments the counter according to the data received on that topic.
  * 
  * \author Baba HADJ SAID
  * \date 24/05/2023
*/




#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <actionlib/client/terminal_state.h>
#include <assignment_2_2022/PlanningAction.h>
#include <actionlib_msgs/GoalStatusArray.h>

#include <sstream>
#include <std_srvs/Trigger.h>



int status; ///< current goal status.
int tmp; ///< holds previous goal status. 
int num_reached_goal=0,num_canceled=0;  ///< reached and cancelled counters.




/**
 * @brief Callback function for updating the reached and cancelled counters.
 *
 * This function is called when a message is received on the topic "/reaching_goal/status". 
 * If the status is 3 (goal reached), it increments the count of reached goals.
 * If the status is 2 (goal canceled), it increments the count of canceled goals.
 * It also keeps track of the previous status value (tmp) to detect changes in status and avoid duplicate publishing of messages.

 * @param msg The received GoalStatusArray message.
 */
void MyCallBack(const actionlib_msgs::GoalStatusArray::ConstPtr& msg){

  if (!msg->status_list.empty()){
  status= msg->status_list[0].status;
  }
  if(status == 3){
    if(status != tmp){
      num_reached_goal++;
    }
  }else if(status==2){
    if(status != tmp){
      num_canceled++;
    }  
  }
  tmp = status;
}

/**
 * 
 * @brief Callback function for handling Trigger service requests.
 *
 * This function is called when a Trigger service is called. 
 * It prints information about the number of goals reached and goals canceled.
 * The function sets the response success field to true, indicating a successful response.
 * 
 * @param req The Trigger service request object.
 * @param res The Trigger service response object.
 * @return True to indicate a successful service call.
 */
bool callback(std_srvs::Trigger::Request& req, std_srvs::Trigger::Response& res)
{
    ROS_INFO("the number of goal reached is %d\nThe number of goal cacelled is %d",num_reached_goal,num_canceled);
    res.success = true;
    return true;
}



int main (int argc, char **argv)
{
	//init a node 
  ros::init(argc, argv, "custom_service");

  ros::NodeHandle nodeHandler;
  //create a subscriber to listen to the status 
  ros::Subscriber sub = nodeHandler.subscribe("/reaching_goal/status", 100, MyCallBack);
  //create the service "my_custom_server_topic" 
  ros::ServiceServer service = nodeHandler.advertiseService("my_custom_server_topic", callback);
  // ros spin to execute callback
  ros::spin();
  return 0;
}

File: src/client_action/script/custom_service.py
#! /usr/local/bin/python
from __future__ import print_function
import rospy
from client_action.srv import my_service,my_serviceResponse
from actionlib_msgs.msg import GoalStatusArray
from std_msgs.msg import String


class My_Service:
    
    statuss = 0
    tmp = 0
    num_reached_goal=0
    num_canceled=0; 
    myString = String()
    def __init__(self):
        self.pub = rospy.Publisher("custom_service_topic",String,queue_size=1000)
        rospy.init_node("custom_service_py")
        self.service = rospy.Service("my_custom_server_py",my_service,self.callback)
        self.subscriber = rospy.Subscriber("/reaching_goal/status",GoalStatusArray,self.MyCallBack,queue_size=100)
    def MyCallBack(self,message):
        if (len(message.status_list) > 0):
            self.statuss = message.status_list[0].status
        
        if(self.statuss==3):
            if self.statuss !=self.tmp:
                self.num_reached_goal+=1
                self.myString.data = f"number of Goal Reached is {self.num_reached_goal}"
                self.pub.publish(self.myString)
        if(self.statuss==2):
            if self.statuss !=self.tmp:
                self.num_canceled+=1
                self.myString.data = f"number of Goal Reached is {self.num_canceled}"
                self.pub.publish(self.myString)
        self.tmp = self.statuss

    def callback(self,req):
        rospy.loginfo(f"the number of goal reached is {self.num_reached_goal}\nThe number of goal cacelled is {self.num_canceled}")
        return my_serviceResponse()
    def main(self):
        
        rospy.loginfo("All Set..")
        rospy.spin()


if __name__ == "__main__":
    My_Service().main()
